Add functions help scrape

kylebarron opened the pull request:
It was pretty quick to write up for functions specifically.

This is an HTML scrape of all names that begin with `f_` in `ado/base/f/`.

Commit:
Add functions help scrape

mcaceresb commented:
So the idea is to have a pre-generated list with HTML help instead of calling Stata for help?

kylebarron commented:
This PR isn't necessarily meant to be merged. Just to attempt an initial scraping. It would be possible to store HTML help instead of loading from Stata, but I don't think that's really a big deal.

But it would be nice to have a current list of keywords, and I think this is the best way to do that.

kylebarron commented:
Also it's a lull in actual work today so I decided to implement function autocomplete.

Commit:
Updates
- Save as a single dict
- Convert HTML to markdown
- Undo Stata's line wrapping of text
- Convert relative links to absolute links

Commit:
Add updated functions.json

kylebarron commented:
pretty!

![peek 2018-09-13 16-31](https://user-images.githubusercontent.com/15164633/45514105-7d3b6780-b772-11e8-911f-ba6e7c62e927.gif)

mcaceresb commented:
Would it make sense to do this instead of fetching help for code inspection? (With the fetch as a fallback)

kylebarron commented:
I don't know. Maybe if I thought it was entirely stable.

kylebarron commented:
I currently convert to Markdown because Atom only takes Markdown (without html tags) as input for the autocomplete descriptions.

mcaceresb commented:
Enclose all the things in `` ``` ``?

By stability you mean because the introspection branch isn't merged yet? I think the idea of the implementation's fine bc it's not a very complicated function. The main thing to iron out is what the behavior ought to be overall.

kylebarron commented:
> Enclose all the things in ```?

Well right now some things are correctly bold and links work, which is nice. It's just that some `_word_` instances aren't correctly turned into _emphasis_ by Marked. I'm not sure why.

> By stability you mean because the introspection branch isn't merged yet? 

No I was referring to the stability of this help file parsing.

kylebarron commented:
I've been working on command help file scraping code... From scraping all the terms under `ado/base/a`, I've found the following list of terms:

```python
['adoupdate',
 'alpha',
 'ameans',
 'anova',
 'anovadef',
 'append',
 'arch',
 'areg',
 'arfima',
 'args',
 'arima',
 'asclogit',
 'asmixlogit',
 'asmprobit',
 'asroprobit',
 'assert',
 'set autotabgraphs']
```

For each of these I have the corresponding HTML, so that I can see what is underlined/bolded/(un)necessary:
```html
<p>
        <b><u>an</u></b><b>ova</b> <a href="/help.cgi?varname"><i>varname</i></a> [<i>termlist</i>] [<a href="/help.cgi?if"><i>if</i></a>] [<a href="/help.cgi?in"><i>in</i></a>] [<a href="/help.cgi?anova#weight"><i>weight</i></a>] [<b>,</b> <i>options</i>]
</p>
```

mcaceresb commented:
I was recommended `getcmds` over in Statalist. Apparently it was written by a StataCorp person, and it basically loops over everything in ado (and the user's local ado folders): 

```stata
net install getcmds, from(http://www.stata.com/users/jpitblado
getcmds using installed.txt, all
```

This makes a list of every installed command, but I think it's using the same idea as your scrapping (while it does list more commands under `a`, not all are useful). It might be good to compare lists; it also doesn't do functions (or, obviously, fetch HTML help).

kylebarron commented:
Interesting. It does show that there's value to looping over files with `.ado` suffixes and not just files with `.sthlp` suffixes.
